test_vector_handlers/src/awses_test_vectors/manifests/mpl_keyring.py

```diff
@@ -0,0 +1,182 @@
+
+This REQUIRES the aws-cryptographic-material-providers library.
+"""
+import json
+import attr
+
+# Ignore missing MPL for pylint, but the MPL is required for this example
+# noqa pylint: disable=import-error
+from aws_cryptography_materialproviderstestvectorkeys.smithygenerated.\
+    aws_cryptography_materialproviderstestvectorkeys.models import (
+        GetKeyDescriptionInput,
+        GetKeyDescriptionOutput,
+        TestVectorKeyringInput,
+    )
+from aws_cryptographic_materialproviders.mpl import AwsCryptographicMaterialProviders
+from aws_cryptographic_materialproviders.mpl.config import MaterialProvidersConfig
+from aws_cryptographic_materialproviders.mpl.references import IKeyring
+from aws_cryptographic_materialproviders.mpl.models import CreateMultiKeyringInput
+
+import _dafny
+import UTF8
+
+# Ignore pylint not being able to read a module that requires the MPL
+# pylint: disable=no-name-in-module
+from awses_test_vectors.internal.mpl.keyvectors_provider import KeyVectorsProvider
+from awses_test_vectors.manifests.keys import KeysManifest  # noqa: disable=F401
+
+
+from .master_key import MasterKeySpec
+
+
+@attr.s
+class KeyringSpec(MasterKeySpec):  # pylint: disable=too-many-instance-attributes
+    """AWS Encryption SDK master key specification utilities.
+
+    Described in AWS Crypto Tools Test Vector Framework features #0003 and #0004.
+
+    :param str type_name: Master key type name
+    :param str key_name: Name of key in keys spec
+    :param str provider_id: Master key provider ID
+    :param str encryption_algorithm: Wrapping key encryption algorithm (required for raw master keys)
+    :param str padding_algorithm: Wrapping key padding algorithm (required for raw master keys)
+    :param str padding_hash: Wrapping key padding hash (required for raw master keys)
+    """
+
+    def keyring(self, keys_uri, mode):
+        # type: (KeysManifest) -> IKeyring
+        """Build a keyring using this specification.
+        :param str keys_uri: Path to the keys manifest
+        """
+        keyvectors = KeyVectorsProvider.get_keyvectors(keys_path=keys_uri)
+
+        # Variable to flag whether we changed anything in weird hack #1.
+        # Signals to weird hack #2 whether it should execute.
+        changed_key_name_from_private_to_public = False
+
+        # Construct the input to KeyVectorsConfig
+        input_kwargs = {
+            "type": self.type_name,
+            "key": self.key_name,
+            "provider-id": self.provider_id,
+            "encryption-algorithm": self.encryption_algorithm,
+
+        }
+        if self.padding_algorithm is not None and self.padding_algorithm != "":
+            input_kwargs["padding-algorithm"] = self.padding_algorithm
+        if self.padding_hash is not None:
+            input_kwargs["padding-hash"] = self.padding_hash
+
+        if input_kwargs["type"] == "raw" \
+                and input_kwargs["encryption-algorithm"] == "rsa":
+            # Weird hack #1:
```

Comment:
Weird hack #1 is also present in Java. https://github.com/aws/aws-encryption-sdk-java/blob/master/src/test/java/com/amazonaws/encryptionsdk/TestVectorGenerator.java#L414-L420

Reply:
Does this mean we have a simplified tests. Eg it specifies the private key to both encrypt and decrypt?

Reply:
The `decrypt-generate` (i.e. `encrypt`) vector manifest specifies a "private" key, but it really should be grabbing the public key.